app/models/user.rb

```diff
@@ -1,6 +1,35 @@
 class User < ActiveRecord::Base
+  mount_uploader :avatar, AvatarUploader
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
-end
+  #:authentication_keys => [:login]
+   attr_accessor :login
+
+  def self.find_for_database_authentication (warden_conditions)
+  # def self.find_first_by_auth_conditions(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
+    else
+      where(conditions).first
+    end
+  end
+
+  validates :username,
+            uniqueness: {
+            case_sensitive: false
```

Comment:
Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.